plugin/plugin.go

```diff
@@ -0,0 +1,96 @@
+package plugin
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/docker/libnetwork/ipamapi"
+	weaveapi "github.com/weaveworks/weave/api"
+	"github.com/weaveworks/weave/common"
+	"github.com/weaveworks/weave/common/docker"
+	weavenet "github.com/weaveworks/weave/net"
+	ipamplugin "github.com/weaveworks/weave/plugin/ipam"
+	netplugin "github.com/weaveworks/weave/plugin/net"
+	"github.com/weaveworks/weave/plugin/skel"
+)
+
+var Log = common.Log
+
+func Start(weaveAPIAddr string, dockerClient *docker.Client, address string, meshAddress string, noDNS bool, isPluginV2 bool) {
+	weave := weaveapi.NewClient(weaveAPIAddr, Log)
+	Log.Info("Waiting for Weave API Server...")
+	weave.WaitAPIServer(30)
+	Log.Info("Finished waiting for Weave API Server")
+
+	if err := run(dockerClient, weave, address, meshAddress, noDNS, isPluginV2); err != nil {
+		Log.Fatal(err)
+	}
+}
+
+func run(dockerClient *docker.Client, weave *weaveapi.Client, address, meshAddress string, noDNS, isPluginV2 bool) error {
+	endChan := make(chan error, 1)
+
+	if address != "" {
+		globalListener, err := listenAndServe(dockerClient, weave, address, endChan, "global", false, noDNS, isPluginV2)
+		if err != nil {
+			return err
+		}
+		defer os.Remove(address)
+		defer globalListener.Close()
+	}
+	if meshAddress != "" {
+		meshListener, err := listenAndServe(dockerClient, weave, meshAddress, endChan, "local", true, noDNS, isPluginV2)
+		if err != nil {
+			return err
+		}
+		defer os.Remove(meshAddress)
+		defer meshListener.Close()
+	}
+
+	statusListener, err := weavenet.ListenUnixSocket("/home/weave/plugin-status.sock")
+	if err != nil {
+		return err
+	}
+	go serveStatus(statusListener)
+
+	return <-endChan
+}
+
+func listenAndServe(dockerClient *docker.Client, weave *weaveapi.Client, address string, endChan chan<- error, scope string, withIpam, noDNS bool, isPluginV2 bool) (net.Listener, error) {
+	name := strings.TrimSuffix(path.Base(address), ".sock")
+	// TODO(mp) fix plugin name, as it is used to determine whether network is our
```

Comment:
Addressed, but in an ugly way:(

The problem is that Docker Plugin V2 does not expose a plugin name, so we need to search for `"net-plugin"` within a driver name of a network. Anyway, suggestions welcome.